server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -341,15 +355,19 @@ public TranslogReader closeIntoReader() throws IOException {
     public TranslogSnapshot newSnapshot() {
         // make sure to acquire the sync lock first, to prevent dead locks with threads calling
         // syncUpTo() , where the sync lock is acquired first, following by the synchronize(this)
+        // After the sync lock we acquire the write lock to avoid deadlocks with threads writing where
+        // the write lock is acquired first followed by synchronize(this).
         synchronized (syncLock) {
-            synchronized (this) {
-                ensureOpen();
-                try {
-                    sync();
-                } catch (IOException e) {
-                    throw new TranslogException(shardId, "exception while syncing before creating a snapshot", e);
+            try (ReleasableLock toClose = writeLock.acquire()) {
+                synchronized (this) {
+                    ensureOpen();
+                    try {
+                        sync();
+                    } catch (IOException e) {
+                        throw new TranslogException(shardId, "exception while syncing before creating a snapshot", e);
+                    }
+                    return super.newSnapshot();
```

Comment:
I think the assert on bufferedOps size and totalOps above can also go here?

Reply:
I added some assertions here based on the sync snapshot. The sync must also have completed successfully to get here. Written offset is annoying because I throws an IOException.